packages/sdk/src/tests/multi/transactions.test.ts

```diff
@@ -0,0 +1,214 @@
+        riverConfig.base.rpcUrl,
+        bobsOtherWallet,
+    )
+    const chainId = riverConfig.base.chainConfig.chainId
+
+    // updated once and shared between tests
+    let bob: SyncAgent
+    let alice: SyncAgent
+    let spaceId: string
+    let defaultChannelId: string
+    // updated before each test
+    let dummyReceipt: ethers.ContractReceipt
+    let dummyReceiptCopy: ethers.ContractReceipt
+
+    beforeAll(async () => {
+        // setup once
+        const log = base_log.extend('beforeAll')
+        log('start')
+
+        await Promise.all([
+            bobIdentity.fundWallet(),
+            aliceIdentity.fundWallet(),
+            bobsOtherWalletProvider.fundWallet(),
+        ])
+
+        bob = await bobIdentity.makeSyncAgent()
+        alice = await aliceIdentity.makeSyncAgent()
+
+        await Promise.all([
+            bob.start(),
+            alice.start(),
+            bob.riverConnection.spaceDapp.walletLink.linkWalletToRootKey(
+                bobIdentity.signer,
+                bobsOtherWallet,
+            ),
+        ])
+
+        // before they can do anything on river, they need to be in a space
+        const { spaceId: sid, defaultChannelId: cid } = await bob.spaces.createSpace(
+            { spaceName: 'BlastOff' },
+            bobIdentity.signer,
+        )
+        spaceId = sid
+        defaultChannelId = cid
+
+        await alice.spaces.joinSpace(spaceId, aliceIdentity.signer)
+        log('bob and alice joined space', spaceId, defaultChannelId)
+
+        const transaction = await bobIdentity.web3Provider.mintMockNFT(riverConfig.base.chainConfig)
+        dummyReceipt = await transaction.wait(2)
+        dummyReceiptCopy = deepCopy(dummyReceipt)
+        expect(dummyReceiptCopy).toEqual(dummyReceipt)
+    })
+
+    afterEach(() => {
+        // for speed we share a receipt between tests, don't modify it
+        expect(dummyReceiptCopy).toEqual(dummyReceipt)
+    })
+
+    test('addEvent', async () => {
+        // a user should be able to upload a transaction that
+        // is from their account or one of their linked accounts and
+        // is valid on chain
+        // add the transaction to the river
+        const transaction = await bobIdentity.web3Provider.mintMockNFT(riverConfig.base.chainConfig)
+        const receipt = await transaction.wait(2)
+        await expect(
+            bob.riverConnection.client!.addTransaction(chainId, receipt),
+        ).resolves.not.toThrow()
+    })
+
+    test('cantAddEventTwice', async () => {
+        const transaction = await bobIdentity.web3Provider.mintMockNFT(riverConfig.base.chainConfig)
+        const receipt = await transaction.wait(2)
+        await expect(
+            bob.riverConnection.client!.addTransaction(chainId, receipt),
+        ).resolves.not.toThrow()
+        // can't add the same transaction twice
+        await expect(bob.riverConnection.client!.addTransaction(chainId, receipt)).rejects.toThrow()
```

Comment:
A nice-to-have for these would be a regex matching the thrown error, `/duplicate/` or something, as an argument to toThrow

Reply:
ah yes i missed this one